### PR TITLE
Added command to `subscribe` and `unsubscribe` to other chains

### LINKED ISSUE #1342: Command to subscribe to other chains
Currently the only way to subscribe to another chain is to submit a GraphQL mutation, but this is an important step in our distributed application deployment process so it should probably have a CLI command.

### PR DESCRIPTION
**Pull request by ashu26jha**
## Motivation

To add command to `subscribe` and `unsubscribe` to other chains, earlier it was only possible through GraphQL mutation, after successfully merge this PR, it can be done through CLI. For example:
`linera subscribe --subscriber $CHAIN_2 --publisher $CHAIN_1 --channel admin`

## Proposal

Added subscribe and unsubscribe in [`linera.rs`](https://github.com/linera-io/linera-protocol/blob/main/linera-service/src/linera.rs) also had to add `clap::ValueEnum` in the `SystemChannel`

Should close #1342

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

**Comment by ashu26jha** [comment]
Build the linera again, also did `linera help-markdown > CLI.md` but it is still adding me to add : ````Possible values: `true`, `false` ````

File: linera-service/src/linera.rs [hunk 1]
```diff
@@ -848,6 +848,36 @@ enum ClientCommand {
         balance: Amount,
     },
 
+    /// Subscribes to a system channel.
+    Subscribe {
+        /// Chain id (must be one of our chains).
+        #[arg(long = "subscriber")]
+        subscriber: ChainId,
```
Comment (ma2bd) [inline linera-service/src/linera.rs:855]: Following other commands, I think we'd like `chain_id: Option<ChainId>` here.

**Comment by afck** [comment]
`CLI.md` must always be auto-generated using:

```
linera help-markdown > CLI.md
```

The text can only be modified by changing the doc strings on the `ClientCommand`s.

**Comment by ashu26jha** [comment]
> `CLI.md` must always be auto-generated using:
> 
> ```
> linera help-markdown > CLI.md
> ```
> 
> The text can only be modified by changing the doc strings on the `ClientCommand`s.

When I tried `cargo run --bin linera -- help-markdown` I am getting the desired output. But I don't get with `linera help-markdown`. I reinstalled linera with :
```
cargo install --path linera-sdk
cargo install --path linera-service
```

**Comment by afck** [comment]
I can't reproduce that:

```
git fetch ashu26jha subscribe-cli
git checkout subscribe-cli 
cargo install --path linera-service
diff CLI.md <(cargo run --bin linera -- help-markdown)
diff CLI.md <(linera -- help-markdown)
```

The last two lines both give me the same output as we're seeing in CI.
Can you double-check whether the `linera` binary you run is really the one Cargo installed? Maybe you set an alias or `PATH` is pointing somewhere else?

**Comment by ashu26jha** [comment]
I guess this explains, thanks!

File: linera-service/src/linera.rs [hunk 2]
```diff
@@ -848,6 +848,36 @@ enum ClientCommand {
         balance: Amount,
     },
 
+    /// Subscribes to a system channel.
+    Subscribe {
+        /// Chain id (must be one of our chains).
+        #[arg(long = "subscriber")]
+        subscriber: Option<ChainId>,
+
+        /// Chain id (must be one of our chains).
```
Comment (ma2bd) [inline linera-service/src/linera.rs:857]: Any owner should be fine but they need to process their inbox and create a block to accept the subscription (for now at least -- we have plans to remove this constraint)

File: linera-service/src/linera.rs [hunk 3]
```diff
@@ -848,6 +848,36 @@ enum ClientCommand {
...
+        #[arg(long = "publisher")]
```
Comment (ma2bd) [inline linera-service/src/linera.rs:858]: nit: I believe you don't need `= "publisher"`

File: linera-service/src/linera.rs [hunk 4]
```diff
@@ -848,6 +848,36 @@ enum ClientCommand {
...
+    /// Unsubscribes from a system channel.
+    Unsubscribe {
+        /// Chain id (must be one of our chains).
+        #[arg(long = "subscriber")]
+        subscriber: Option<ChainId>,
+
+        /// Chain id (must be one of our chains).
+        #[arg(long = "publisher")]
+        publisher: Option<ChainId>,
+
```
Comment (ma2bd) [inline linera-service/src/linera.rs:875]: see other comments

File: linera-service/src/linera.rs [hunk 5]
```diff
@@ -1545,6 +1575,61 @@ impl Runnable for Job {
...
+                                .subscribe_to_published_bytecodes(publisher_chainid)
+                                .await
+                        }
+                        None => Err(ChainClientError::InternalError("Incorrect chain ID")),
+                    },
+                };
+                context.update_and_save_wallet(&mut chain_client).await;
+                let subscribe = result.context("Failed to subscribe")?;
+                let time_total = time_start.elapsed().as_micros();
+                info!("Subscription confirmed after {} us", time_total);
```
Comment (ma2bd) [inline linera-service/src/linera.rs:1600]: We're in the process of switching to `millis`.

**Review by ma2bd — APPROVED** [review]
Thanks for the changes. Just a few more details

File: linera-service/src/linera.rs (hunk 2 above)
Comment (ashu26jha) [reply linera-service/src/linera.rs:857]: Thanks for reviewing it, can you elaborate bit more?

File: linera-service/src/linera.rs (hunk 5 above)
Comment (ashu26jha) [reply linera-service/src/linera.rs:1600]: Okay 👍

File: linera-service/src/linera.rs (hunk 2 above)
Comment (ma2bd) [reply linera-service/src/linera.rs:857]: I'm just saying that the comment "(must be one of our chains)" does not apply

File: linera-service/src/linera.rs (hunk 5 above)
Comment (jvff) [inline linera-service/src/linera.rs:1600]: I'd suggest using `Duration`'s `Debug` formatter to print out the value with the unit:

```suggestion
                let time_total = time_start.elapsed();
                info!("Subscription confirmed after {time_total:.2?}");
```

Comment (afck) [reply linera-service/src/linera.rs:1600]: The plan is to use milliseconds in most places. They're probably a reasonable choice here.

File: linera-service/src/linera.rs (hunk 3 above)
Comment (ashu26jha) [reply linera-service/src/linera.rs:858]: Cool, btw should we also remove `= "subscriber"`

File: linera-service/src/linera.rs (hunk 2 above)
Comment (ashu26jha) [reply linera-service/src/linera.rs:857]: Ok

File: linera-service/src/linera.rs (hunk 5 above)
Comment (ashu26jha) [reply linera-service/src/linera.rs:1600]: Cool

File: linera-service/src/linera.rs (hunk 3 above)
Comment (afck) [reply linera-service/src/linera.rs:858]: Yes, it's only needed if you want the argument to have a different name than the field:
https://docs.rs/clap/latest/clap/_derive/index.html#arg-attributes

Comment (ashu26jha) [reply linera-service/src/linera.rs:858]: Cool thanks! That helps

**Comment by jvff** [comment]
Replaced by #1564, which is just this PR rebased on top of main and with the final review suggestions applied.

**Pull request closed without merging**